src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,230 @@
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsThisParameter,
+				(SymbolKind.Method, SymbolKind.Parameter) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsParameter :
+					DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsParameter,
+				(SymbolKind.NamedType, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Method) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsMethodReturnType,
+				(SymbolKind.TypeParameter, SymbolKind.Parameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsParameter,
+				(SymbolKind.TypeParameter, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsGenericParameter,
+				_ => throw new NotImplementedException ()
+			};
+
+		static string[] GetDiagnosticArguments (ISymbol source, ISymbol target, string missingAnnotations)
+		{
+			var args = new List<string> ();
+			args.AddRange (GetDiagnosticArguments (target));
+			args.AddRange (GetDiagnosticArguments (source));
+			args.Add (missingAnnotations);
+			return args.ToArray ();
+		}
+
+		static IEnumerable<string> GetDiagnosticArguments (ISymbol symbol)
+		{
+			var args = new List<string> ();
+			args.AddRange (symbol.Kind switch {
+				SymbolKind.Parameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				SymbolKind.NamedType => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Field => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Method => new string[] { symbol.GetDisplayName () },
+				SymbolKind.TypeParameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				_ => throw new NotImplementedException ($"Unsupported source or target symbol {symbol}.")
+			});
+
+			return args;
+		}
+
+		static ISymbol? TryGetSymbolFromOperation (IOperation? operation) =>
+			operation switch {
+				IArgumentOperation argument => TryGetSymbolFromOperation (argument.Value),
+				IAssignmentOperation assignment => TryGetSymbolFromOperation (assignment.Value),
+				IConversionOperation conversion => conversion.Type,
+				IInstanceReferenceOperation instanceReference => instanceReference.Type,
+				IInvocationOperation invocation => invocation.TargetMethod,
+				IMemberReferenceOperation memberReference => memberReference.Member,
+				IParameterReferenceOperation parameterReference => parameterReference.Parameter,
+				IReturnOperation returnOp => TryGetSymbolFromOperation (returnOp.ReturnedValue),
+				ITypeOfOperation typeOf => typeOf.TypeOperand,
```

Comment:
Will the TypeOperand be a NamedType? I imagine this could cause ambiguities with the instanceReference case above. Could we just remove this case for now (I don't see where it is used)?

Reply:
This will always be a TypeParameter and is needed for cases where there's a `typeof` expression. I might be missing something here but I don't see why/how should we remove it.